File: desktop/protocol.py
"""JSON protocol helpers. Small and explicit."""

from typing import Any


def ok(data: Any) -> dict[str, Any]:
    return {"ok": True, "data": data}


def fail(code: str, message: str, recover: str = "") -> dict[str, Any]:
    err: dict[str, Any] = {"code": code, "message": message}
    if recover:
        err["recover"] = recover
    return {"ok": False, "error": err}


def partial(data: dict[str, Any], code: str, message: str,
            recover: str = "") -> dict[str, Any]:
    """Failure envelope that also carries partial data (batch receipts).

    Standard shape for interrupted batches: ok=false with error, plus data
    holding status/last_completed/receipts. Executed events cannot be
    undone; receipts report exactly what ran.
    """
    env = fail(code, message, recover)
    env["data"] = data
    return env

File: desktop/rust/src/protocol.rs
use serde_json::{Value, json};

pub const COMMANDS: &[&str] = &[
    "observe",
    "list_apps",
    "list_windows",
    "active-window",
    "focus_app",
    "launch",
    "move",
    "click",
    "double-click",
    "right_click",
    "drag",
    "scroll",
    "type",
    "click_type",
    "press",
    "hotkey",
    "wait",
    "batch",
    "stop",
    "doctor",
];

pub fn ok(data: Value) -> Value {
    json!({"ok": true, "data": data})
}

pub fn fail(code: &str, message: &str, recover: &str) -> Value {
    let mut error = json!({"code": code, "message": message});
    if !recover.is_empty() {
        error["recover"] = json!(recover);
    }
    json!({"ok": false, "error": error})
}

pub fn partial(data: Value, code: &str, message: &str, recover: &str) -> Value {
    let mut envelope = fail(code, message, recover);
    envelope["data"] = data;
    envelope
}
